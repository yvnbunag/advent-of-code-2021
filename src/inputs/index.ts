import { readFileSync } from 'fs'

type Numbers = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 0
type Days = `${Numbers}${Numbers}` | Numbers
type FileName = `day-${Days}-${Numbers}`

export function getInput(filename: FileName): string {
  return readFileSync(`${__dirname}/${filename}.txt`, { encoding: 'utf8' })
}
