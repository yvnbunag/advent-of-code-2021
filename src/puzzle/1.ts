export function answer() {
  return 1
}
