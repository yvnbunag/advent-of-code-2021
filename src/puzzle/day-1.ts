import { parseInputToList } from '~/puzzle/utils'

export function calculateIncrease(report: string): number {
  return parseInputToList(report)
    .map(Number)
    .reduce((count, currentValue, currentIndex, original) => {
      if (!currentIndex) return 0

      if (currentValue > original[currentIndex - 1]) return count + 1

      return count
    }, 0)
}

export function calculateGroupedIncrease(report: string) {
  const values = parseInputToList(report).map(Number)
  const windowedValues = values
    .reduce(
      (collection, value, index, original) => {
        return [
          ...collection,
          original
            .slice(index, index + 3)
            .reduce((value, current) => value + current,0),
        ]
      },
    [] as Array<number>,
    )
    .slice(0, -2)

  return calculateIncrease(windowedValues.join('\n'))
}
