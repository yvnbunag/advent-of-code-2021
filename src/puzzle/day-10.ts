import { parseInputToList } from '~/puzzle/utils'

import { add, subtract } from '~/puzzle/utils/number'

type First = unknown

type Second = unknown

const map = {
  '{': '}',
  '(': ')',
  '[': ']',
  '<': '>',
}

const vmap = {
  ')': 3,
  ']': 57,
  '}': 1197,
  '>': 25137,
}

const vvmap = {
  ')': 1,
  ']': 2,
  '}': 3,
  '>': 4,
}

function isOpening(v: string): boolean {
  return Object.keys(map).includes(v)
}

class BrokenError extends Error {}
class IncompleteError extends Error {}

export function first(input: string): First {
  const data = parseInputToList(input)
  const errors: Array<Error> = []

  for (const line of data) {
    try {
      const opening: Array<string> = []

      for (const char of line.split('')) {
        if (isOpening(char)) {
          opening.push(char)
        }

        else {
          const pair = opening.pop() as keyof typeof map
          if (map[pair] !== char) {
            throw new BrokenError(char)
          }
        }
      }

      if (opening.length) throw new IncompleteError()
    } catch (err: any) {
      errors.push(err)
    }
  }

  const broken = errors
    .filter((error) => error instanceof BrokenError)
    .map(({ message }) => message)
    // @ts-expect-error asd
    .map((key) => vmap[key])
    .reduce(add, 0)

  return broken
}

export function second(input: string): Second {
  const data = parseInputToList(input)
  const errors: Array<Error> = []

  for (const line of data) {
    try {
      const opening: Array<string> = []

      for (const char of line.split('')) {
        if (isOpening(char)) {
          opening.push(char)
        }

        else {
          const pair = opening.pop() as keyof typeof map
          if (map[pair] !== char) {
            throw new BrokenError(char)
          }
        }
      }

      if (opening.length) throw new IncompleteError(opening.join(''))
    } catch (err: any) {
      errors.push(err)
    }
  }

  const incomplete = errors
    .filter((error) => error instanceof IncompleteError)
    .map(({ message }) => message)
    .map((v) => v.split('').reverse()
    // @ts-expect-error asd
      .map((v) => map[v]))

  const answer = incomplete.map((v) => {

    return v.reduce((q, w) => {
      // @ts-expect-error asd
      const ww = vvmap[w]

      return (q * 5) + ww
    }, 0)
  })

  return answer.sort(subtract)[Math.floor(answer.length / 2)]
}
