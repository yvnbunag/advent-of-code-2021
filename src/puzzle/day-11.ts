/* eslint-disable @typescript-eslint/naming-convention */
import { parseInputToMatrix } from '~/puzzle/utils'

type Position = {
  y: number,
  x: number,
}

type Flash = {
  isPeak: boolean,
  center: Position,
}

type PeakFlash = Flash & { isPeak: true }

class Octopus {
  static PEAK_ENERGY = 9
  static LOW_ENERGY = 0

  readonly position: Position

  constructor(
    y: Position['y'],
    x: Position['x'],
    private _energy: number = Octopus.LOW_ENERGY,
  ) {
    this.position = { y, x }
  }

  get energy() { return this._energy }

  step(): Flash {
    const energy = this.incrementEnergy()
    const isPeak = energy === Octopus.PEAK_ENERGY + 1

    return { isPeak, center: this.position }
  }

  settle(): Octopus {
    if (this.energy > Octopus.PEAK_ENERGY) this.resetEnergy()

    return this
  }

  private incrementEnergy() {
    this._energy = this._energy + 1

    return this._energy
  }

  private resetEnergy() { this._energy = Octopus.LOW_ENERGY }

  static step(octopus: Octopus): ReturnType<Octopus['step']> {
    return octopus.step()
  }

  static settle(octopus: Octopus): ReturnType<Octopus['settle']> {
    return octopus.settle()
  }

  static getPosition(octopus: Octopus): Position {
    return octopus.position
  }
}

type OctopusRow = Array<Octopus>

type OctopusMatrix = Array<OctopusRow>

function toOctopusMatrix(row: Array<number>, y: number): OctopusRow {
  return row.map((energy: number, x: number) => {
    return new Octopus(y, x, energy)
  })
}

function parseToOctopusMatrix(input: string): OctopusMatrix {
  return parseInputToMatrix(input)
    .map((row) => row.map(Number))
    .map(toOctopusMatrix)
}

function getOctopusFromMatrix(
  matrix: OctopusMatrix,
  { y, x }: Position,
): Octopus {
  return matrix[y][x]
}

function getAdjacentOctopus(
  position: Position,
  matrix: OctopusMatrix,
): Array<Octopus> {
  const { y, x } = position
  const top = y > 0 && { y: y - 1, x }
  const bottom = matrix.length > y + 1 && { y: y + 1, x }
  const left = x > 0 && { y, x: x - 1 }
  const right = matrix[y].length > x + 1 && { y, x: x + 1 }
  const topLeft = top && left && { y: top.y, x: left.x }
  const topRight = top && right && { y: top.y, x: right.x }
  const bottomLeft = bottom && left && { y: bottom.y, x: left.x }
  const bottomRight = bottom && right && { y: bottom.y, x: right.x }

  const positions = [
    top,
    bottom,
    left,
    right,
    topLeft,
    topRight,
    bottomLeft,
    bottomRight,
  ].filter((position) => position !== false) as Array<Position>

  return positions.map((position) => getOctopusFromMatrix(matrix, position))
}

function isPeakFlash(flash: Flash): flash is PeakFlash {
  return flash.isPeak
}

function isSynchronized(matrix: OctopusMatrix): boolean {
  return !matrix.some((row) => row.some((octopus) => octopus.energy > 0))
}

export function calculateFlashes(input: string, steps: number) {
  function spreadFlash(
    matrix: OctopusMatrix,
    flashes: Array<PeakFlash>,
    count = 0,
  ): number {
    if (flashes.length === 0) return count

    const [{ center }, ...succeeding] = flashes
    const adjacent = getAdjacentOctopus(center, matrix)
      .map(Octopus.step)
      .filter(isPeakFlash)

    return spreadFlash(matrix, [...succeeding, ...adjacent], count + 1)
  }

  function calculate(
    matrix: OctopusMatrix,
    steps: number,
    count = 0,
  ): number {
    if (steps === 0) return count

    const flashes = matrix
      .flatMap((row) => row.map(Octopus.step))
      .filter(isPeakFlash)
    const spreadCount = spreadFlash(matrix, flashes)
    const settledMatrix = matrix.map((row) => row.map(Octopus.settle))

    return calculate(settledMatrix, steps - 1, count + spreadCount)
  }

  const matrix = parseToOctopusMatrix(input)

  return calculate(matrix, steps)
}

export function calculateSynchronization(input: string) {
  function spreadFlash(
    matrix: OctopusMatrix,
    flashes: Array<PeakFlash>,
  ): void {
    if (flashes.length === 0) return

    const [{ center }, ...succeeding] = flashes
    const adjacent = getAdjacentOctopus(center, matrix)
      .map(Octopus.step)
      .filter(isPeakFlash)

    return spreadFlash(matrix, [...succeeding, ...adjacent])
  }

  function calculate(
    matrix: OctopusMatrix,
    until: (matrix: OctopusMatrix)=> boolean,
    steps = 1,
  ): number {
    const flashes = matrix
      .flatMap((row) => row.map(Octopus.step))
      .filter(isPeakFlash)

    spreadFlash(matrix, flashes)

    const settledMatrix = matrix.map((row) => row.map(Octopus.settle))

    if (until(settledMatrix)) return steps

    return calculate(settledMatrix, until, steps + 1)
  }

  const matrix = parseToOctopusMatrix(input)

  return calculate(matrix, isSynchronized)
}
