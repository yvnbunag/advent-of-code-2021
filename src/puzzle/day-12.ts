import { parseInputToList } from '~/puzzle/utils'
import { getLastEntry } from '~/puzzle/utils/list'
import { not, isLowerCase } from '~/puzzle/utils/predicate'

type Point = string

type From = Point

type To = Point

type Path = Array<Point>

type Paths = Array<Path>

type Connection = [From, To]

type Connections = Array<Connection>

type PathMap = Record<Point, Array<Point>>

const START: Point = 'start'

const END: Point = 'end'

function parseInputToConnections(
  input: string,
  start: Point,
  end: Point,
): Connections {
  const paths = parseInputToList(input)
    .map((rawPath) => rawPath.split('-').slice(0, 2) as Connection)
  const returnPaths = paths
    .filter(([from, to]) => from !== start && to !== end)
    .map(([from, to]) => [to, from] as Connection)

  return [...paths, ...returnPaths]
}

function toPathMap(connections: Connections): PathMap {
  return connections.reduce((pathMap, [from, to]) => {
    return {
      ...pathMap,
      [from]: [...(pathMap[from] || []), to],
    }
  }, {} as PathMap)
}

function createHasEnded(end: Point) {
  return (path: Path) => getLastEntry(path) === end
}

function createHasUniqueLowercasePoints(overflow = 0) {
  return (path: Path) => {
    const lowercase = path.filter(isLowerCase)
    const uniqueLowercase = [...new Set(lowercase)]

    return lowercase.length < uniqueLowercase.length + 1 + overflow
  }
}

export function findDistinctPaths(input: string, smallRepeat = 0): Paths {
  const hasUniqueLowercasePoints = createHasUniqueLowercasePoints(smallRepeat)
  const hasEnded = createHasEnded(END)

  function find(
    pathMap: PathMap,
    start: Point,
    pending: Paths = pathMap[start].map((to) => [start, to]),
    ended: Paths = [],
  ): Paths {
    if (!pending.length) return ended

    const branches = pending.reduce((accumulated, path) => {
      return [
        ...accumulated,
        ...(pathMap[getLastEntry(path)] || [])
          .filter((point) => point !== start)
          .map((point) => [...path, point].filter(Boolean))
          .filter(hasUniqueLowercasePoints),
      ]
    }, [] as Paths)
    const endBranches = branches.filter(hasEnded)
    const pendingBranches = branches.filter(not(hasEnded))

    return find(pathMap, start, pendingBranches, [...ended, ...endBranches])
  }

  const connections = parseInputToConnections(input, START, END)
  const pathMap = toPathMap(connections)

  return find(pathMap, START)
}

export function findSimpleDistinctPaths(input: string) {
  return findDistinctPaths(input, 0).length
}

export function findComplexDistinctPaths(input: string) {
  return findDistinctPaths(input, 1).length
}
