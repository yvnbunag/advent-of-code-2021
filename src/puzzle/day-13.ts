import { parseInputToList } from '~/puzzle/utils'

import { splitBy } from '~/puzzle/utils/string'
import { not } from '~/puzzle/utils/predicate'
import { getHighestValue } from '~/puzzle/utils/list'
import { deepCopy } from '~/puzzle/utils/matrix'

// @TODO Verify solution

type Pixel = boolean

type Row = Array<Pixel>

type Paper = Array<Row>

type Axis = 'x' | 'y'

type RawInstruction = `fold along ${Axis}=${number}`

type RawLocation = `${number},${number}`

type Instruction = {
  axis: Axis,
  position: number,
}

type Manual = {
  paper: Paper,
  instructions: Array<Instruction>,
}

function parseLocationsToPaper(rawLocation: Array<RawLocation>): Paper {
  const locations = rawLocation.map(splitBy(','))
    .map((location) => location.map(Number))
  const width = getHighestValue(locations.map(([x]) => x)) + 1
  const height = getHighestValue(locations.map(([,y]) => y)) + 1
  const row = Array(width).fill(false)
  const paper = Array(height).fill([])
    .map(() => [...row])

  return locations.reduce((currentPaper, [x, y]) => {
    currentPaper[y][x] = true

    return currentPaper
  }, deepCopy(paper))
}

function parseInstruction(rawInstruction: RawInstruction) {
  const [,,data] = rawInstruction.split(' ')
  const [axis, position] = data.split('=') as [Axis, number]

  return { axis, position: Number(position) }
}

function isLocation(value: string): value is RawLocation {
  return value.includes(',')
}

const isRawInstruction = (() => {
  type IsInstruction = (value: string)=> value is RawInstruction

  return not(isLocation) as IsInstruction
})()

function parseInputToManual(input: string): Manual {
  const data = parseInputToList(input)
  const paper = parseLocationsToPaper(data.filter(isLocation))
  const instructions: Array<Instruction> = data
    .filter(isRawInstruction)
    .map(parseInstruction)

  return { paper, instructions }
}

export function plot(paper: Paper) {
  return paper
    .map((row) => row.map((pixel) => pixel ? '#' : '.').join(''))
    .join('\n')
}

function foldLeft(paper: Paper, position: number): Paper {
  const left = paper.map((row) => row.slice(0, position))
  const right = paper.map((row) => row.slice(position + 1))
  const foldedRight = right.map((row) => [...row].reverse())

  return left.map((row, y) => {
    return row.map((pixel, x) => {
      return pixel || foldedRight?.[y]?.[x] || false
    })
  })
}

function foldUp(paper: Paper, position: number): Paper {
  const upper = paper.slice(0, position)
  const lower = paper.slice(position + 1)
  const foldedLower = [...lower].reverse()

  return upper.map((row, y) => {
    return row.map((pixel, x) => {
      return pixel || foldedLower?.[y]?.[x] || false
    })
  })
}

export function fold(input: string, limit = Infinity): Paper {
  function recursiveFold(
    paper: Paper,
    instructions: Array<Instruction>,
    limit = Infinity,
  ): Paper {
    if (!instructions.length) return paper
    if (!limit) return paper

    const [{ axis, position }, ...nextInstructions] = instructions

    if (axis === 'y') {
      return recursiveFold(foldUp(paper, position), nextInstructions, limit - 1)
    }

    return recursiveFold(foldLeft(paper, position), nextInstructions, limit - 1)
  }

  const { paper, instructions } = parseInputToManual(input)

  return recursiveFold(paper, instructions, limit)
}

export function countPixels(paper: Paper): number {
  return paper.flatMap((value) => value).filter(Boolean).length
}
