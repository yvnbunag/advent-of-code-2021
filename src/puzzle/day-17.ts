import { parseInputToList } from '~/puzzle/utils'

export type Trajectory = {
  x: number,
  y: number,
}

export type Position = {
  x: number,
  y: number,
}

type Range = [number, number]

export type TargetArea = {
  x: Range,
  y: Range,
}

type Probe = {
  trajectory: Trajectory,
  position: Position,
}

type History = Array<Probe>

function parseToTargetArea(input: string) {
  const x = input.match(/x=(-*\d+)..(-*\d+)/)?.slice(1, 3)
    .map(Number)
  const y = input.match(/y=(-*\d+)..(-*\d+)/)?.slice(1, 3)
    .map(Number)

  return { x, y } as TargetArea
}

export function shoot(
  probe: Probe,
  until: (history: History)=> boolean = () => false,
): History {
  const history = [probe]
  const currentPosition: Position = { ...probe.position }
  const currentTrajectory: Trajectory = { ...probe.trajectory }

  while (!until(history)) {
    if (currentTrajectory.x > 0) {
      currentPosition.x = currentPosition.x + currentTrajectory.x--
    }

    currentPosition.y = currentPosition.y + currentTrajectory.y--

    history.push({
      position: { ...currentPosition },
      trajectory: { ...currentTrajectory },
    })
  }

  return history
}

export function isWithinTarget(area: TargetArea, { x, y }: Position) {
  const withinX = x >= area.x[0] && x <= area.x[1]
  const withinY = y >= area.y[0] && y <= area.y[1]

  return withinX && withinY
}

export function isBeyondTarget(area: TargetArea, { x, y }: Position) {
  const beyondX = x > area.x[1]
  const beyondY = y < area.y[0]

  return beyondX || beyondY
}

export function first(input: string) {
  const data = parseToTargetArea(input)
  const initial: Position = { x: 0, y: 0 }

  console.log(data)

  return 0
}

export function second(input: string) {
  const data = parseInputToList(input)

  return 0
}
