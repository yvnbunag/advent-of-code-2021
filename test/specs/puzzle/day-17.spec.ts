import { shoot, isWithinTarget, isBeyondTarget, first, second } from '~/puzzle/day-17'
import { getInput } from '~/inputs'

import type { TargetArea, Position, Trajectory } from '~/puzzle/day-17'

describe('Methods', () => {
  test('shoot', () => {
    const position: Position = { x: 0, y: 0 }
    const trajectory: Trajectory = { x:7, y: 2 }
    const probe = { position, trajectory }
    const history = shoot(probe, ({ length }) => length > 7 )
    const [
      p1, p2, p3, p4, p5, p6, p7, p8,
    ] = history

    expect(p1).toMatchSnapshot()
    expect(p2).toMatchSnapshot()
    expect(p3).toMatchSnapshot()
    expect(p4).toMatchSnapshot()
    expect(p5).toMatchSnapshot()
    expect(p6).toMatchSnapshot()
    expect(p7).toMatchSnapshot()
    expect(p8).toMatchSnapshot()
  })

  describe('isWithinTarget', () => {
    test('Inside 1', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 28,
        y: -7,
      }
      expect(isWithinTarget(target, position)).toBe(true)
    })

    test('Outside 1', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 27,
        y: -3,
      }
      expect(isWithinTarget(target, position)).toBe(false)
    })

    test('Inside 2', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 22,
        y: -9,
      }
      expect(isWithinTarget(target, position)).toBe(true)
    })

    test('Outside 2', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 22,
        y: -4,
      }
      expect(isWithinTarget(target, position)).toBe(false)
    })
  })

  describe('isBeyondTarget', () => {
    test('before 1', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 27,
        y: -3,
      }
      expect(isBeyondTarget(target, position)).toBe(false)
    })

    test('within 1', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 28,
        y: -7,
      }
      expect(isBeyondTarget(target, position)).toBe(false)
    })

    test('beyond 1', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 28,
        y: -12,
      }
      expect(isBeyondTarget(target, position)).toBe(true)
    })

    test('before 2', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 24,
        y: -3,
      }
      expect(isBeyondTarget(target, position)).toBe(false)
    })

    test('within 2', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 30,
        y: -6,
      }
      expect(isBeyondTarget(target, position)).toBe(false)
    })

    test('beyond 2', () => {
      const target = {
        x: [20, 30],
        y: [-10, -5],
      } as TargetArea
      const position: Position = {
        x: 35,
        y: -10,
      }
      expect(isBeyondTarget(target, position)).toBe(true)
    })
  })
})

describe('Part 1 ', () => {
  it('should answer example', () => {
    const data = `target area: x=20..30, y=-10..-5`

    // expect(first(data)).toBe(0)
  })

  it('should answer puzzle', () => {
    // const data = getInput('day-17')

    // expect(first(data)).toMatchInlineSnapshot(`0`)
  })
})

describe('Part 2', () => {
  it('should answer example', () => {
    const data = ``

    expect(second(data)).toBe(0)
  })

  it('should answer puzzle', () => {
    const data = getInput('day-17')

    expect(second(data)).toMatchInlineSnapshot(`0`)
  })
})
