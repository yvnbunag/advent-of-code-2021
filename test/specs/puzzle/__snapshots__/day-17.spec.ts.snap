// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Methods shoot 1`] = `
Object {
  "position": Object {
    "x": 0,
    "y": 0,
  },
  "trajectory": Object {
    "x": 7,
    "y": 2,
  },
}
`;

exports[`Methods shoot 2`] = `
Object {
  "position": Object {
    "x": 7,
    "y": 2,
  },
  "trajectory": Object {
    "x": 6,
    "y": 1,
  },
}
`;

exports[`Methods shoot 3`] = `
Object {
  "position": Object {
    "x": 13,
    "y": 3,
  },
  "trajectory": Object {
    "x": 5,
    "y": 0,
  },
}
`;

exports[`Methods shoot 4`] = `
Object {
  "position": Object {
    "x": 18,
    "y": 3,
  },
  "trajectory": Object {
    "x": 4,
    "y": -1,
  },
}
`;

exports[`Methods shoot 5`] = `
Object {
  "position": Object {
    "x": 22,
    "y": 2,
  },
  "trajectory": Object {
    "x": 3,
    "y": -2,
  },
}
`;

exports[`Methods shoot 6`] = `
Object {
  "position": Object {
    "x": 25,
    "y": 0,
  },
  "trajectory": Object {
    "x": 2,
    "y": -3,
  },
}
`;

exports[`Methods shoot 7`] = `
Object {
  "position": Object {
    "x": 27,
    "y": -3,
  },
  "trajectory": Object {
    "x": 1,
    "y": -4,
  },
}
`;

exports[`Methods shoot 8`] = `
Object {
  "position": Object {
    "x": 28,
    "y": -7,
  },
  "trajectory": Object {
    "x": 0,
    "y": -5,
  },
}
`;
